phoenix-core/src/main/java/org/apache/phoenix/compile/IndexColumnExpressionCompiler.java

```diff
@@ -0,0 +1,311 @@
+        for (PColumn column : context.getCurrentTable().getTable().getColumns()) {
+            if (column.getExpressionStr() != null) {
+                Expression expression = null;
+                try {
+                    ParseNode parseNode = SQLParser.parseCondition(column.getExpressionStr());
+                    expression = getExpression(parseNode);
+                } catch (SQLException e) {
+                    throw new RuntimeException(e);
+                }
+                expressionMap.put(column.getExpressionStr(), expression);
+            }
+        }
+    }
+
+    /**
+     * Returns the compiled expression
+     * 
+     * @param node
+     *            data table parse node
+     */
+    private Expression getExpression(ParseNode node) throws SQLException {
+        PTable indexTable = this.context.getCurrentTable().getTable();
+        NamedTableNode dataTableNode = NamedTableNode.create(null, TableName.create(indexTable.getParentSchemaName()
+                .getString(), indexTable.getParentTableName().getString()), Collections.<ColumnDef> emptyList());
+        ColumnResolver resolver = FromCompiler.getResolver(dataTableNode, this.context.getConnection());
+        StatementContext context = new StatementContext(this.context.getStatement(), resolver, new Scan(),
+                new SequenceManager(this.context.getStatement()));
+        ExpressionCompiler expressionCompiler = new ExpressionCompiler(context);
+        return node.accept(expressionCompiler);
+    }
+
+    /**
+     * @return true if current table is an index and there is an expression that matches the given node
+     */
+    private boolean matchesIndexedExpression(ParseNode node) throws SQLException {
+        if (context.getCurrentTable().getTable().getType() != PTableType.INDEX) { 
+            return false; 
+        }
+        DataTableNodeRewriter statementRewriter = new DataTableNodeRewriter();
+        ParseNode dataTableParseNode = node.accept(statementRewriter);
+        Expression dataTableExpression = getExpression(dataTableParseNode);
+        return expressionMap.containsKey(dataTableExpression.toString());
+    }
+
+    private Expression convertAndVisitParseNode(ParseNode node) throws SQLException {
+        DataTableNodeRewriter statementRewriter = new DataTableNodeRewriter();
+        ParseNode dataTableParseNode = node.accept(statementRewriter);
+        Expression expression = getExpression(dataTableParseNode);
+        ColumnParseNode columnParseNode = new ColumnParseNode(null, IndexUtil.getIndexColumnName(null,
+                expression.toString()), null);
+        PDataType expressionType = expression.getDataType();
+        PDataType indexColType = IndexUtil.getIndexColumnDataType(expression.isNullable(), expression.getDataType());
+        // if data type of expression is different from the index column data type, cast the to expression type
+        if (indexColType != expressionType) {
```

Comment:
I added this so that we don't need to specify a CAST in order to use the index (similar to the check in IndexStatementRewriter)